### PR TITLE
Poly discrete rotate

### PR DESCRIPTION
**Pull request by remi-or**
## Motivation
Please describe the motivation of this PR and the goal you want to achieve through this PR.
This PR adds the PolyDiscreteRotate transform, a transform that functions like PolyRandomRotate but allows the user to specify a list of angles to choose from. It also fixes two typos in the PolyRandomRotate docstring. 

## Modification
The added PolyDiscreteTransform is added to the transforms.py file, inheriting from the existing PolyRandomRotate. The parent transform bears no change but two fixed typos in its docstring.

## BC-breaking (Optional)
No BC breaking.

## Use cases (Optional)
The PolyDiscreteRotate transform is used (under the name RandomRotateRbox with 'value' mode) in the [FCOSR repo](https://github.com/lzh420202/FCOSR/blob/ca5f40ebf11aa4dbf40eabba326fe1acbf5dc590/configs/fcosrbox/dataset/dota10_single.py). It allows for more control on the Rotation augmentation by changing the continuous distribution the angle is chosen from to a discrete set. This also allows to control the resulting size of the rotated image if no resizing is done afterwards.

## Checklist

1. All pre-commit tools were run except markdownlint which bugs at installation, but no changes were made to .md files.
2. PolyDiscreteRotate was addid with a unit test similar to PolyRandomRotate and passed it.
3. The documentation was added along with the class.

### CODE REVIEW

**Comment by yangxue0827** [comment]
Thank you for your contributions. Please follow https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md to set pre-commit and fix the lint error.

**Comment by jbwang1997** [comment]
Thanks for your contributions. I wonder if we can directly extend the feature of `PolyRandomRotate`. Look like most codes of `PolyDiscreteRotate` is the same as `PolyRandomRotate`.

**Comment by jbwang1997** [comment]
Thanks for your contributions. I wonder if we can directly extend the feature of `PolyRandomRotate`. Look like most codes of `PolyDiscreteRotate` is the same as `PolyRandomRotate`.

**Comment by remi-or** [comment]
Ok, then I think the simplest solution is to add a list of `base_angles` as an argument to `PolyRandomRotate` which defaults to `[0]`, and to change the rotation angle from `(-angles_range, +angles_range)` to `base_angle + (-angles_range, +angles_range)` where `base_angle` is chosen uniformly in `base_angles`.  

That way: 
- default behavior is unchanged
- no BC problem (except if the transform was created w/out `build_from_cfg` and positional arguments, which doesn't seems to be the case anywhere) 
- the transform if flexible, allowing for discrete angles (w/ `angles_range = 0`), random angles in a given range, or a mix of the two.

In my opinion, it's better than a `mode` argument that would be set to `"values"` or `"range"`
I can draft the PR if you agree.

**Comment by jbwang1997** [comment]
> Ok, then I think the simplest solution is to add a list of `base_angles` as an argument to `PolyRandomRotate` which defaults to `[0]`, and to change the rotation angle from `(-angles_range, +angles_range)` to `base_angle + (-angles_range, +angles_range)` where `base_angle` is chosen uniformly in `base_angles`.
> 
> That way:
> 
> * default behavior is unchanged
> * no BC problem (except if the transform was created w/out `build_from_cfg` and positional arguments, which doesn't seems to be the case anywhere)
> * the transform if flexible, allowing for discrete angles (w/ `angles_range = 0`), random angles in a given range, or a mix of the two.
> 
> In my opinion, it's better than a `mode` argument that would be set to `"values"` or `"range"` I can draft the PR if you agree.

Great! You can first draft this feature.

File: mmrotate/datasets/pipelines/transforms.py [hunk 1]
```diff
@@ -177,8 +183,10 @@ def __call__(self, results):
             results['rotate'] = False
             angle = 0
         else:
-            angle = 2 * self.angles_range * np.random.rand() - \
-                    self.angles_range
+            i_angle = np.random.randint(len(self.base_angles))
+            base_angle = self.base_angles[i_angle]
```
Comment (jbwang1997) [inline mmrotate/datasets/pipelines/transforms.py:187]: base_angle = np.random.choice(self.base_angles)

File: mmrotate/datasets/pipelines/transforms.py [hunk 2]
```diff
@@ -93,28 +93,34 @@ def bbox_flip(self, bboxes, img_shape, direction):
 
 @ROTATED_PIPELINES.register_module()
 class PolyRandomRotate(object):
-    """Rotate img & bbox.
+    """Rotate img & bbox with angle theta = base_angle + random_angle where:
+        - base angle is chosen uniformly in the list of base_angles
+        - random_angle is a random float in (-angles_range, +angles_range)
```
[the remark this answers is not in the document]
Comment (yangxue0827) [reply mmrotate/datasets/pipelines/transforms.py:98]: Actually, I don't yet understand what `base_angles` is used for. Doesn't it also randomly return the rotation angle in a continuous value range after adding `base_angles`? The only difference is the range of angle values.

[the remark this answers is not in the document]
Comment (jbwang1997) [reply mmrotate/datasets/pipelines/transforms.py:98]: In my understanding, when we set `base_angles` as a list of angle values and `angle_ranges` as 0, this pipeline can discretely select the angle value in `base_angles`.

[the remark this answers is not in the document]
Comment (jbwang1997) [reply mmrotate/datasets/pipelines/transforms.py:98]: By combining `base_angles` and `angle_ranges`, we can deal with all cases of discrete and continuous angle rotation.

[the remark this answers is not in the document]
Comment (yangxue0827) [reply mmrotate/datasets/pipelines/transforms.py:98]: I prefer to use the `mode` parameter to control whether `value` mode or `range` mode is selected.

[the remark this answers is not in the document]
Comment (jbwang1997) [reply mmrotate/datasets/pipelines/transforms.py:98]: You are right. `mode` pattern may be more intuitive.

**Comment by jbwang1997** [comment]
Hello @remi-or. After discussion, we consider controlling with `mode` may be more intuitive for users. Now implementation may be hard to get the usage directly from argument names.

File: mmrotate/datasets/pipelines/transforms.py [hunk 3]
```diff
@@ -97,25 +97,46 @@ class PolyRandomRotate(object):
...
+                raise TypeError("PolyRandomRotate with mode 'range' \
+                                  expects angle_range to be an int.")
+        elif mode == 'value':
+            if not (isinstance(angles_range, list)
+                    and isinstance(angles_range[0], int)):
+                raise TypeError("PolyRandomRotate with mode 'value' expects \
+                                  angle_range to be a non-empty list of int.")
+        else:
+            raise ValueError(f"Unknown mode for PolyRandomRange: {mode}. \
+                               Mode is supposed to be 'range' or 'value'.")
```
Comment (jbwang1997) [inline mmrotate/datasets/pipelines/transforms.py:138]: We can direct limit the mode by `assert`.
assert mode in ['range', 'vaule], f"`model` is supposed to be 'range' or 'value', but get {model}."

File: mmrotate/datasets/pipelines/transforms.py [hunk 4]
```diff
@@ -97,25 +97,46 @@ class PolyRandomRotate(object):
...
                  version='le90'):
         self.rotate_ratio = rotate_ratio
         self.auto_bound = auto_bound
+        if mode == 'range':
+            if not isinstance(angles_range, int):
+                raise TypeError("PolyRandomRotate with mode 'range' \
+                                  expects angle_range to be an int.")
+        elif mode == 'value':
+            if not (isinstance(angles_range, list)
+                    and isinstance(angles_range[0], int)):
```
Comment (jbwang1997) [inline mmrotate/datasets/pipelines/transforms.py:133]: mmcv.is_seq_of(angles_range, int)

File: mmrotate/datasets/pipelines/transforms.py [hunk 5]
```diff
@@ -177,9 +198,14 @@ def __call__(self, results):
...
-            angle = 2 * self.angles_range * np.random.rand() - \
-                    self.angles_range
             results['rotate'] = True
+            if self.mode == 'range':
+                angle = self.angles_range * (2 * np.random.rand() - 1)
+            elif self.mode == 'value':
+                i = np.random.randint(len(self.angles_range))
+                angle = self.angles_range[i]
+            else:
+                raise ValueError(f'Unknown mode: {self.mode}')
```
Comment (jbwang1997) [inline mmrotate/datasets/pipelines/transforms.py:208]: We have limited `mode` in __init__. Here is redundant.

File: mmrotate/datasets/pipelines/transforms.py (hunk 4 above)
Comment (remi-or) [reply mmrotate/datasets/pipelines/transforms.py]: didn't know about that one, thanks!

**Review by jbwang1997 — APPROVED** [review]
LGTM

**Review by yangxue0827 — APPROVED** [review]
LGTM

**Pull request merged**